File: verifiers/go/verify/verify.go
//go:build ignore

// AlgoVoi substrate-pqc — Go verifier.
//
// Reads an AP2 PQC v0 artefact JSON, recomputes the JCS canonical bytes,
// confirms the canonical SHA-256 byte-anchor, and verifies each declared
// signature scheme.
//
// Supported schemes: ES256, Ed25519, ML-DSA-65
// (Falcon-1024 not supported — no pure-Go lib without CGo.)
//
// Prints one PASS/FAIL line per scheme; exits 0 iff all pass.
//
// Usage:
//
//	go run verifiers/go/verify/verify.go <artefact.json>
package main

import (
	"crypto/ecdsa"
	"crypto/ed25519"
	"crypto/sha256"
	"crypto/x509"
	"encoding/asn1"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"math/big"
	"os"

	"github.com/cloudflare/circl/sign/mldsa/mldsa65"
	jsoncanonicalizer "github.com/cyberphone/json-canonicalization/go/src/webpki.org/jsoncanonicalizer"
)

func b64d(s string) ([]byte, error) {
	return base64.StdEncoding.DecodeString(s)
}

func jcsBytes(v interface{}) ([]byte, error) {
	raw, err := json.Marshal(v)
	if err != nil {
		return nil, err
	}
	return jsoncanonicalizer.Transform(raw)
}

// ecdsaDSASig holds r and s for DER decoding.
type ecdsaDSASig struct {
	R, S *big.Int
}

func verifyES256(canonical []byte, sig map[string]interface{}) error {
	// Load public key — accept DER (publicKeyDer) or raw 65-byte uncompressed (publicKey_b64).
	var pub *ecdsa.PublicKey
	if derB64, ok := sig["publicKeyDer"].(string); ok {
		derBytes, err := b64d(derB64)
		if err != nil {
			return fmt.Errorf("publicKeyDer base64: %w", err)
		}
		k, err := x509.ParsePKIXPublicKey(derBytes)
		if err != nil {
			return fmt.Errorf("parse PKIX public key: %w", err)
		}
		var ok2 bool
		pub, ok2 = k.(*ecdsa.PublicKey)
		if !ok2 {
			return fmt.Errorf("not an ECDSA public key")
		}
	} else if rawB64, ok := sig["publicKey_b64"].(string); ok {
		rawBytes, err := b64d(rawB64)
		if err != nil {
			return fmt.Errorf("publicKey_b64 base64: %w", err)
		}
		if len(rawBytes) != 65 || rawBytes[0] != 0x04 {
			return fmt.Errorf("expected 65-byte uncompressed P-256 point")
		}
		// Wrap as SPKI DER for x509 parsing.
		spki := buildP256SPKI(rawBytes)
		k, err := x509.ParsePKIXPublicKey(spki)
		if err != nil {
			return fmt.Errorf("parse P-256 raw point: %w", err)
		}
		var ok2 bool
		pub, ok2 = k.(*ecdsa.PublicKey)
		if !ok2 {
			return fmt.Errorf("not an ECDSA public key")
		}
	} else {
		return fmt.Errorf("no publicKeyDer or publicKey_b64 field")
	}

	// Load signature — accept DER (signature_der) or compact 64-byte r||s.
	var sigDer []byte
	if derB64, ok := sig["signature_der"].(string); ok {
		var err error
		sigDer, err = b64d(derB64)
		if err != nil {
			return fmt.Errorf("signature_der base64: %w", err)
		}
	} else {
		for _, key := range []string{"signature_b64", "signature_compact_b64"} {
			if compB64, ok := sig[key].(string); ok {
				compact, err := b64d(compB64)
				if err != nil {
					return fmt.Errorf("%s base64: %w", key, err)
				}
				if len(compact) == 64 {
					r := new(big.Int).SetBytes(compact[:32])
					s := new(big.Int).SetBytes(compact[32:])
					sigDer, err = asn1.Marshal(ecdsaDSASig{R: r, S: s})
					if err != nil {
						return fmt.Errorf("compact->DER: %w", err)
					}
				} else {
					sigDer = compact
				}
				break
			}
		}
	}
	if sigDer == nil {
		return fmt.Errorf("no signature_der / signature_b64 / signature_compact_b64 field")
	}

	// ES256 hashes internally: hash canonical with SHA-256.
	digest := sha256.Sum256(canonical)
	var dsaSig ecdsaDSASig
	if _, err := asn1.Unmarshal(sigDer, &dsaSig); err != nil {
		return fmt.Errorf("unmarshal DER signature: %w", err)
	}
	if !ecdsa.Verify(pub, digest[:], dsaSig.R, dsaSig.S) {
		return fmt.Errorf("signature invalid")
	}
	return nil
}

func verifyEd25519(canonical []byte, sig map[string]interface{}) error {
	pubB64, ok := sig["publicKey_b64"].(string)
	if !ok {
		return fmt.Errorf("missing publicKey_b64")
	}
	sigB64, ok := sig["signature_b64"].(string)
	if !ok {
		return fmt.Errorf("missing signature_b64")
	}
	pubBytes, err := b64d(pubB64)
	if err != nil {
		return fmt.Errorf("publicKey_b64 base64: %w", err)
	}
	sigBytes, err := b64d(sigB64)
	if err != nil {
		return fmt.Errorf("signature_b64 base64: %w", err)
	}
	if len(pubBytes) != ed25519.PublicKeySize {
		return fmt.Errorf("expected %d-byte Ed25519 public key, got %d", ed25519.PublicKeySize, len(pubBytes))
	}
	if !ed25519.Verify(ed25519.PublicKey(pubBytes), canonical, sigBytes) {
		return fmt.Errorf("signature invalid")
	}
	return nil
}

func verifyMLDSA65(canonical []byte, sig map[string]interface{}) error {
	pubB64, ok := sig["publicKey_b64"].(string)
	if !ok {
		return fmt.Errorf("missing publicKey_b64")
	}
	sigB64, ok := sig["signature_b64"].(string)
	if !ok {
		return fmt.Errorf("missing signature_b64")
	}
	pubBytes, err := b64d(pubB64)
	if err != nil {
		return fmt.Errorf("publicKey_b64 base64: %w", err)
	}
	sigBytes, err := b64d(sigB64)
	if err != nil {
		return fmt.Errorf("signature_b64 base64: %w", err)
	}
	var pub mldsa65.PublicKey
	if err := pub.UnmarshalBinary(pubBytes); err != nil {
		return fmt.Errorf("unmarshal ML-DSA-65 public key: %w", err)
	}
	if !mldsa65.Verify(&pub, canonical, nil, sigBytes) {
		return fmt.Errorf("signature invalid")
	}
	return nil
}

// buildP256SPKI wraps an uncompressed P-256 point (65 bytes) into a
// SubjectPublicKeyInfo DER structure compatible with x509.ParsePKIXPublicKey.
func buildP256SPKI(raw []byte) []byte {
	// Algorithm identifier OID sequence for EC P-256:
	//   SEQUENCE { OID ecPublicKey, OID prime256v1 }
	// = 30 13 06 07 2a 86 48 ce 3d 02 01 06 08 2a 86 48 ce 3d 03 01 07
	algID := []byte{0x30, 0x13,
		0x06, 0x07, 0x2a, 0x86, 0x48, 0xce, 0x3d, 0x02, 0x01,
		0x06, 0x08, 0x2a, 0x86, 0x48, 0xce, 0x3d, 0x03, 0x01, 0x07}
	// BIT STRING: unused bits byte (0x00) + raw point
	bs := append([]byte{0x00}, raw...)
	bitStr := make([]byte, 0, 2+len(bs))
	bitStr = append(bitStr, 0x03)
	bitStr = append(bitStr, byte(len(bs)))
	bitStr = append(bitStr, bs...)
	// SEQUENCE { algID, bitStr }
	inner := append(algID, bitStr...)
	spki := make([]byte, 0, 2+len(inner))
	spki = append(spki, 0x30)
	spki = append(spki, byte(len(inner)))
	spki = append(spki, inner...)
	return spki
}

func main() {
	if len(os.Args) < 2 {
		fmt.Fprintln(os.Stderr, "usage: verify <artefact.json>")
		os.Exit(2)
	}
	artefactPath := os.Args[1]
	data, err := os.ReadFile(artefactPath)
	if err != nil {
		fmt.Fprintf(os.Stderr, "read artefact: %v\n", err)
		os.Exit(1)
	}

	var artefact map[string]interface{}
	if err := json.Unmarshal(data, &artefact); err != nil {
		fmt.Fprintf(os.Stderr, "parse JSON: %v\n", err)
		os.Exit(1)
	}

	// Recompute JCS canonical bytes from mandate_body.
	mandateBody, ok := artefact["mandate_body"]
	if !ok {
		mandateBody = artefact["payload"]
	}
	if mandateBody == nil {
		fmt.Fprintln(os.Stderr, "artefact missing mandate_body / payload")
		os.Exit(1)
	}

	canonical, err := jcsBytes(mandateBody)
	if err != nil {
		fmt.Fprintf(os.Stderr, "JCS error: %v\n", err)
		os.Exit(1)
	}

	digest := sha256.Sum256(canonical)
	gotSHA := "sha256:" + hex.EncodeToString(digest[:])
	expectedSHA, _ := artefact["expected_canonical_sha256"].(string)
	shaOK := gotSHA == expectedSHA
	fmt.Printf("canonical_sha_ok=%v\n", shaOK)
	if !shaOK {
		fmt.Fprintf(os.Stderr, "canonical SHA mismatch: got %s want %s\n", gotSHA, expectedSHA)
	}

	sigsRaw, _ := artefact["signatures"].(map[string]interface{})
	allPass := shaOK

	for alg, sigRaw := range sigsRaw {
		sig, ok := sigRaw.(map[string]interface{})
		if !ok {
			fmt.Printf("FAIL: %s: not a JSON object\n", alg)
			allPass = false
			continue
		}
		var verifyErr error
		switch alg {
		case "ES256":
			verifyErr = verifyES256(canonical, sig)
		case "Ed25519":
			verifyErr = verifyEd25519(canonical, sig)
		case "ML-DSA-65":
			verifyErr = verifyMLDSA65(canonical, sig)
		case "Falcon-1024":
			fmt.Printf("SKIP: Falcon-1024: no pure-Go verifier (no CGo)\n")
			continue
		default:
			fmt.Printf("SKIP: %s: unknown scheme\n", alg)
			continue
		}
		if verifyErr != nil {
			fmt.Printf("FAIL: %s: %v\n", alg, verifyErr)
			allPass = false
		} else {
			fmt.Printf("PASS: %s\n", alg)
		}
	}

	if !allPass {
		os.Exit(1)
	}
}

File: verifiers/rust/src/verify.rs
//! Rust verifier: reads an AP2 PQC v0 artefact and verifies all declared signature schemes.

use base64::{engine::general_purpose::STANDARD as B64, Engine as _};
use ed25519_dalek::{Signature as EdSig, VerifyingKey as Ed25519VerifyKey};
use ml_dsa::{MlDsa65, Signature as MlSig, VerifyingKey as MlVerifyKey};
use p256::ecdsa::{DerSignature, VerifyingKey as P256VerifyKey, signature::Verifier as _};
use pqcrypto_falcon::falcon1024;
use pqcrypto_traits::sign::{DetachedSignature as _, PublicKey as _};
use serde_json::Value;
use sha2::{Digest, Sha256};
use std::{fs, process};

pub fn run(artefact_path: &str) {
    let data = fs::read_to_string(artefact_path).unwrap_or_else(|e| {
        eprintln!("read artefact: {}", e);
        process::exit(1);
    });
    let artefact: Value = serde_json::from_str(&data).unwrap_or_else(|e| {
        eprintln!("parse JSON: {}", e);
        process::exit(1);
    });

    // Get mandate_body (or payload fallback).
    let body = artefact
        .get("mandate_body")
        .or_else(|| artefact.get("payload"))
        .cloned()
        .unwrap_or_else(|| {
            eprintln!("artefact missing mandate_body / payload");
            process::exit(1);
        });

    // Recompute JCS canonical bytes.
    let canonical = serde_jcs::to_vec(&body).unwrap_or_else(|e| {
        eprintln!("JCS error: {}", e);
        process::exit(1);
    });

    let digest = Sha256::digest(&canonical);
    let got_sha = format!("sha256:{}", hex::encode(digest.as_slice()));
    let expected_sha = artefact["expected_canonical_sha256"]
        .as_str()
        .unwrap_or("");
    let sha_ok = got_sha == expected_sha;
    println!("canonical_sha_ok={}", sha_ok);
    if !sha_ok {
        eprintln!(
            "canonical SHA mismatch: got {} want {}",
            got_sha, expected_sha
        );
    }

    let sigs = match artefact.get("signatures").and_then(|v| v.as_object()) {
        Some(s) => s.clone(),
        None => {
            eprintln!("no signatures field");
            process::exit(1);
        }
    };

    let mut all_pass = sha_ok;

    for (alg, sig_val) in &sigs {
        let sig = match sig_val.as_object() {
            Some(o) => o,
            None => {
                println!("FAIL: {}: not a JSON object", alg);
                all_pass = false;
                continue;
            }
        };

        match alg.as_str() {
            "ES256" => match verify_es256(&canonical, sig) {
                Ok(()) => println!("PASS: ES256"),
                Err(e) => {
                    println!("FAIL: ES256: {}", e);
                    all_pass = false;
                }
            },
            "Ed25519" => match verify_ed25519(&canonical, sig) {
                Ok(()) => println!("PASS: Ed25519"),
                Err(e) => {
                    println!("FAIL: Ed25519: {}", e);
                    all_pass = false;
                }
            },
            "ML-DSA-65" => match verify_mldsa65(&canonical, sig) {
                Ok(()) => println!("PASS: ML-DSA-65"),
                Err(e) => {
                    println!("FAIL: ML-DSA-65: {}", e);
                    all_pass = false;
                }
            },
            "Falcon-1024" => match verify_falcon1024(&canonical, sig) {
                Ok(()) => println!("PASS: Falcon-1024"),
                Err(e) => {
                    println!("FAIL: Falcon-1024: {}", e);
                    all_pass = false;
                }
            },
            other => {
                println!("SKIP: {}: unknown scheme", other);
            }
        }
    }

    if !all_pass {
        process::exit(1);
    }
}

fn b64d(s: &str) -> Result<Vec<u8>, String> {
    B64.decode(s).map_err(|e| format!("base64 decode: {}", e))
}

fn get_str<'a>(sig: &'a serde_json::Map<String, Value>, key: &str) -> Result<&'a str, String> {
    sig.get(key)
        .and_then(|v| v.as_str())
        .ok_or_else(|| format!("missing field: {}", key))
}

fn verify_es256(
    canonical: &[u8],
    sig: &serde_json::Map<String, Value>,
) -> Result<(), String> {
    use p256::pkcs8::DecodePublicKey;

    // Load public key — accept DER (publicKeyDer) or raw 65-byte uncompressed (publicKey_b64).
    let pub_key: P256VerifyKey = if let Some(der_b64) = sig
        .get("publicKeyDer")
        .and_then(|v| v.as_str())
    {
        let der = b64d(der_b64)?;
        P256VerifyKey::from_public_key_der(&der)
            .map_err(|e| format!("parse DER public key: {}", e))?
    } else if let Some(raw_b64) = sig.get("publicKey_b64").and_then(|v| v.as_str()) {
        let raw = b64d(raw_b64)?;
        P256VerifyKey::from_sec1_bytes(&raw)
            .map_err(|e| format!("parse SEC1 public key: {}", e))?
    } else {
        return Err("no publicKeyDer or publicKey_b64".to_string());
    };

    // Load signature — accept DER (signature_der) or compact 64-byte r||s.
    let sig_bytes: Vec<u8> = if let Some(der_b64) = sig
        .get("signature_der")
        .and_then(|v| v.as_str())
    {
        b64d(der_b64)?
    } else {
        let compact_b64 = sig
            .get("signature_b64")
            .or_else(|| sig.get("signature_compact_b64"))
            .and_then(|v| v.as_str())
            .ok_or("no signature_der / signature_b64 / signature_compact_b64")?;
        let compact = b64d(compact_b64)?;
        if compact.len() == 64 {
            compact_to_der(&compact)?
        } else {
            compact
        }
    };

    // ES256 hashes with SHA-256 internally (p256 ecdsa verifier).
    let ecdsa_sig = DerSignature::try_from(sig_bytes.as_slice())
        .map_err(|e| format!("parse DER signature: {}", e))?;
    pub_key
        .verify(canonical, &ecdsa_sig)
        .map_err(|e| format!("ES256 verify failed: {}", e))
}

fn compact_to_der(compact: &[u8]) -> Result<Vec<u8>, String> {
    if compact.len() != 64 {
        return Err(format!(
            "expected 64-byte compact signature, got {}",
            compact.len()
        ));
    }
    let r = &compact[..32];
    let s = &compact[32..];
    let r_der = der_int(r);
    let s_der = der_int(s);
    let inner_len = r_der.len() + s_der.len();
    let mut out = Vec::with_capacity(2 + inner_len);
    out.push(0x30u8); // SEQUENCE
    out.push(inner_len as u8);
    out.extend_from_slice(&r_der);
    out.extend_from_slice(&s_der);
    Ok(out)
}

fn der_int(bytes: &[u8]) -> Vec<u8> {
    let trimmed: Vec<u8> = bytes.iter().skip_while(|&&b| b == 0).cloned().collect();
    let need_pad = trimmed.first().map_or(false, |&b| b & 0x80 != 0);
    let content_len = trimmed.len() + if need_pad { 1 } else { 0 };
    let mut out = vec![0x02u8, content_len as u8];
    if need_pad {
        out.push(0x00);
    }
    out.extend_from_slice(&trimmed);
    out
}

fn verify_ed25519(
    canonical: &[u8],
    sig: &serde_json::Map<String, Value>,
) -> Result<(), String> {
    use ed25519_dalek::Verifier;

    let pub_b64 = get_str(sig, "publicKey_b64")?;
    let sig_b64 = get_str(sig, "signature_b64")?;
    let pub_bytes = b64d(pub_b64)?;
    let sig_bytes = b64d(sig_b64)?;

    let pub_arr: [u8; 32] = pub_bytes
        .try_into()
        .map_err(|_| "expected 32-byte Ed25519 public key".to_string())?;
    let sig_arr: [u8; 64] = sig_bytes
        .try_into()
        .map_err(|_| "expected 64-byte Ed25519 signature".to_string())?;

    let vk = Ed25519VerifyKey::from_bytes(&pub_arr)
        .map_err(|e| format!("Ed25519 public key parse: {}", e))?;
    let signature = EdSig::from_bytes(&sig_arr);
    vk.verify(canonical, &signature)
        .map_err(|e| format!("Ed25519 verify failed: {}", e))
}

fn verify_mldsa65(
    canonical: &[u8],
    sig: &serde_json::Map<String, Value>,
) -> Result<(), String> {
    use ml_dsa::{KeyInit, Verifier};

    let pub_b64 = get_str(sig, "publicKey_b64")?;
    let sig_b64 = get_str(sig, "signature_b64")?;
    let pub_bytes = b64d(pub_b64)?;
    let sig_bytes = b64d(sig_b64)?;

    // Build VerifyingKey from bytes using new_from_slice (from KeyInit trait).
    let vk = MlVerifyKey::<MlDsa65>::new_from_slice(&pub_bytes)
        .map_err(|_| format!("ML-DSA-65 public key wrong size (got {} bytes)", pub_bytes.len()))?;

    // Decode signature using TryFrom<&[u8]>.
    let ml_sig = MlSig::<MlDsa65>::try_from(sig_bytes.as_slice())
        .map_err(|_| format!("ML-DSA-65 signature decode failed (got {} bytes)", sig_bytes.len()))?;

    vk.verify(canonical, &ml_sig)
        .map_err(|e| format!("ML-DSA-65 verify failed: {}", e))
}

fn verify_falcon1024(
    canonical: &[u8],
    sig: &serde_json::Map<String, Value>,
) -> Result<(), String> {
    let pub_b64 = get_str(sig, "publicKey_b64")?;
    let sig_b64 = get_str(sig, "signature_b64")?;
    let pub_bytes = b64d(pub_b64)?;
    let sig_bytes = b64d(sig_b64)?;

    let pk = falcon1024::PublicKey::from_bytes(&pub_bytes)
        .map_err(|e| format!("Falcon-1024 public key parse: {}", e))?;
    let dsig = falcon1024::DetachedSignature::from_bytes(&sig_bytes)
        .map_err(|e| format!("Falcon-1024 signature parse: {}", e))?;
    falcon1024::verify_detached_signature(&dsig, canonical, &pk)
        .map_err(|e| format!("Falcon-1024 verify failed: {}", e))
}
